base_layer/core/src/mining/miner.rs

```diff
@@ -134,23 +151,35 @@ impl<B: BlockchainBackend> Miner<B> {
     /// restarts/starts to mine.
     pub async fn mine(mut self) {
         let flag = self.received_new_block_flag.clone();
-        let mut block_event = self.node_interface.clone().get_block_event_stream_fused();
+        let block_event = self.node_interface.clone().get_block_event_stream_fused();
+        let state_event = self.state_change_event_rx.take().expect("todo").fuse();
         let t_miner = self.mining().fuse();
+
+        pin_mut!(block_event);
+        pin_mut!(state_event);
         pin_mut!(t_miner);
         loop {
             futures::select! {
-                msg = block_event.select_next_some() => {
-                    match (*msg).clone() {
-                        BlockEvent::Verified((_, result)) => {
-                            if result == BlockAddResult::Ok{
-                                flag.store(true, Ordering::Relaxed);
-                            }
-                        },
-                        _ => (),
-                    }
+            msg = block_event.select_next_some() => {
+                match (*msg).clone() {
+                    BlockEvent::Verified((_, result)) => {
+                    if result == BlockAddResult::Ok{
+                        flag.store(true, Ordering::Relaxed);
+                    };
                 },
-                (_) = t_miner => break
-            }
+                _ => (),
+                }
+            },
+            msg1 = state_event.select_next_some() => {
```

Comment:
can just be `msg` here doesn't need to be unique.

Reply:
agreed, was made unique as part of debugging.